### PR TITLE
Revised keyVaultSecretExpiryNonRbac

### PR DESCRIPTION
**Pull request by AkhtarAmir**

### CODE REVIEW

File: exports.js [hunk 1]
```diff
@@ -1062,7 +1062,8 @@ module.exports = {
         'sslCertificateAutoRenewal'     : require(__dirname + '/plugins/azure/keyvaults/sslCertificateAutoRenewal.js'),
         'manageKeyAccessAndPermissions' : require(__dirname + '/plugins/azure/keyvaults/manageKeyAccessAndPermissions.js'),
         'rsaCertificateKeySize'         : require(__dirname + '/plugins/azure/keyvaults/rsaCertificateKeySize.js'),
-        'keyVaultSecretExpiry'          : require(__dirname + '/plugins/azure/keyvaults/keyVaultSecretExpiry.js'),
+        'keyVaultSecretExpiry': require(__dirname + '/plugins/azure/keyvaults/keyVaultSecretExpiry.js'),
```
Comment (fatima99s) [inline exports.js:1065]:
```suggestion
        'keyVaultSecretExpiry          :': require(__dirname + '/plugins/azure/keyvaults/keyVaultSecretExpiry.js'),
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'Key Vault Secret Expiry Non RBAC',
+    category: 'Key Vaults',
+    domain: 'Application Integration',
+    severity: 'High',
+    description: 'Proactively check for Key Vault secrets expiry date and rotate them before expiry date is reached.',
```
Comment (fatima99s) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:9]:
```suggestion
    description: 'Proactively check for Non RBAC Key Vault secrets expiry date and rotate them before expiry date is reached.',
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    more_info: 'After the expiry date has reached for Key Vault secret, it cannot be used for storing sensitive and confidential data such as passwords and database connection strings anymore.',
+    recommended_action: 'Ensure that Key Vault secrets are rotated before they get expired.',
+    link: 'https://learn.microsoft.com/en-us/azure/secret-vault/about-secrets-secrets-and-certificates',
+    apis: ['vaults:list', 'vaults:getSecrets'],
+    settings: {
+        key_vault_secret_expiry_fail: {
```
Comment (fatima99s) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:15]:
```suggestion
        non_rbac_key_vault_secret_expiry_fail: {
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }, 
+    realtime_triggers: ['microsoftkeyvault:vaults:write', 'microsoftkeyvault:vaults:delete'],
+
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var locations = helpers.locations(settings.govcloud);
+        var config = {
+            key_vault_secret_expiry_fail: parseInt(settings.key_vault_secret_expiry_fail || this.settings.key_vault_secret_expiry_fail.default)
```
Comment (fatima99s) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:29]:
```suggestion
            key_vault_secret_expiry_fail: parseInt(settings.non_rbac_key_vault_secret_expiry_fail || this.settings.non_rbackey_vault_secret_expiry_fail.default)
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+            if (!vaults.data.length) {
+                helpers.addResult(results, 0, 'No Key Vaults found', location);
+                return rcb();
+            }
+
+            vaults.data.forEach(function(vault) {
+                // Check if vault is non-RBAC
+                if (vault.properties && vault.properties.enableRbacAuthorization) {
+                    helpers.addResult(results, 0,
```
Comment (fatima99s) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:51]: should we not just skip the key vault that have rbac enabled?

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                    return;
+                }
+
+                var secrets = helpers.addSource(cache, source,
+                    ['vaults', 'getSecrets', location, vault.id]);
+
+                if (!secrets || secrets.err || !secrets.data) {
+                    helpers.addResult(results, 3, 'Unable to query for Key Vault secrets: ' + helpers.addError(secrets), location, vault.id);
+                } else if (!secrets.data.length) {
+                    helpers.addResult(results, 0, 'No Key Vault secrets found', location, vault.id);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:60]:
```suggestion
                    helpers.addResult(results, 0, 'No Key Vault secrets found in non RBAC vault', location, vault.id);
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                } else {
+                    secrets.data.forEach(function(secret) {
+                        var secretName = secret.id.substring(secret.id.lastIndexOf('/') + 1);
+                        var secretId = `${vault.id}/secrets/${secretName}`;
+
+                        if (!secret.attributes || !secret.attributes.enabled) {
+                            helpers.addResult(results, 0, 'Secret is not enabled', location, secretId);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:67]:
```suggestion
                            helpers.addResult(results, 0, 'Secret is not enabled in non RBAC vault', location, secretId);
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                        } else if (secret.attributes && (secret.attributes.exp || secret.attributes.expiry)) {
+                            let attributes = secret.attributes;
+                            let secretExpiry = attributes.exp ? attributes.exp * 1000 : attributes.expiry;
+                            let difference = Math.round((new Date(secretExpiry).getTime() - (new Date).getTime())/(24*60*60*1000));
+                            if (difference > config.key_vault_secret_expiry_fail) {
+                                helpers.addResult(results, 0,
+                                    `Secret expires in ${difference} days`, location, secretId);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:74]:
```suggestion
                                    `Secret in non RBAC vault expires in ${difference} days`, location, secretId);
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                            } else if (difference > 0){
+                                helpers.addResult(results, 2,
+                                    `Secret expires in ${difference} days`, location, secretId);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:77]:
```suggestion
                                    `Secret in non RBAC vault expires in ${difference} days`, location, secretId);
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                            } else {
+                                helpers.addResult(results, 2,
+                                    `Secret expired ${Math.abs(difference)} days ago`, location, secretId);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:80]:
```suggestion
                                    `Secret in non RBAC vault expired ${Math.abs(difference)} days ago`, location, secretId);
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+            };
+
+            auth.run(createCache(null, [], {}), {}, callback);
+        });
+
+        it('should give passing result if secret expiration is not enabled in non-RBAC vault', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('Secret expiration is not enabled');
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js:165]:
```suggestion
                expect(results[0].message).to.include('Secret expiration is not enabled in non RBAC vault');
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+            };
+
+            auth.run(createCache(null, [listKeyVaults[0]], getSecrets[3]), {}, callback);
+        });
+
+        it('should give failing result if secret expires within failure expiry date', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('Secret expires');
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js:201]:
```suggestion
                expect(results[0].message).to.include('Secret in non RBAC vault expires');
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+            };
+
+            auth.run(createCache(null, [listKeyVaults[0]], getSecrets[2]), { key_vault_secret_expiry_fail: '40' }, callback);
+        });
+
+        it('should give passing result if key is disabled', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('Secret is not enabled');
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.spec.js:213]:
```suggestion
                expect(results[0].message).to.include('Secret in non RBAC vault is not enabled');
```

File: plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                            }
+                        } else {
+                            helpers.addResult(results, 0,
+                                'Secret expiration is not enabled', location, secretId);
```
Comment (alphadev4) [inline plugins/azure/keyvaults/keyVaultSecretExpiryNonRbac.js:84]:
```suggestion
                                'Secret expiration is not enabled in non RBAC vault', location, secretId);
```

**Review by alphadev4 — APPROVED** [review]

**Pull request merged**